include/pmacc/identifier/value_identifier.hpp

```diff
@@ -47,15 +46,32 @@
  *      `length()` or `length_`
  * @{
  */
-#define value_identifier(in_type, name, in_default)                                                                   \
+
+/** @param ... must be a device function/lambda
+ *
+ * @attention: getValue() is only callable from a pure device function e.g.
+ *
+ * @code{.cpp}
+ * [] ALPAKA_FN_ACC() { return IdProvider<simDim>::getNewId();}
+ * @endcode
+ */
+#define value_identifier_func(in_type, name, ...)                                                                     \
     identifier(                                                                                                       \
-        name, using type = in_type; HDINLINE static type getValue()                                                   \
-        { return in_default; } static std::string getName() { return std::string(#name); })
+        name, using type = in_type; DINLINE static type getValue()                                                    \
+        {                                                                                                             \
+            auto const func = __VA_ARGS__;                                                                            \
+            return func();                                                                                            \
+        } static std::string getName() { return std::string(#name); })
```

Comment:
Just because it's a macro, do we have to throw away all formatting rules and not even have linebreaks around function definitions?

Reply:
Wow! That is a severe limitation of `clang-format`, if it can't handle this properly. (Although it's probably not quite trivial to do.) I've found [SkipMacroDefinitionBody](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#skipmacrodefinitionbody) but that's a bit too broad for my taste.

Reply:
Since we try to avoid large macros this is a rare case that we are affected by the problem.